### PR TITLE
feat(encoding): add json/stream.ts

### LINKED ISSUE #2208: Proposal: Web stream based jsonlines parser
Would you like to add stream based [jsonlines](https://jsonlines.org/) parser as suggested in discord?

https://discord.com/channels/684898665143206084/712010403302866974/972924800852951040

```
{"some":"thing"}
{"foo":17,"bar":false,"quux":true}
{"may":{"include":"nested","objects":["and","arrays"]}}
```

```ts
import { JSONLinesParseStream } from "...";

const { body } = await fetch("...");
const readable = body!
  .pipeThrough(new TextDecoderStream())
  .pipeThrough(new JSONLinesParseStream());
for await (const data of readable) {
  console.log(data);
}
```

Implementations can be copied from https://deno.land/x/jsonlines@v1.1.0.

### PR DESCRIPTION
**Pull request by ayame113**
close #2208

- `JSONLinesParseStream ` parses [JSON lines](https://jsonlines.org/), [NDJSON](http://ndjson.org/) and [JSON Text Sequences](https://datatracker.ietf.org/doc/html/rfc7464).
- `JSONLinesStringifyStream` stringifies [JSON lines](https://jsonlines.org/), [NDJSON](http://ndjson.org/) and [JSON Text Sequences](https://datatracker.ietf.org/doc/html/rfc7464).
- `ConcatenatedJSONParseStream` parses [Concatenated JSON](https://en.wikipedia.org/wiki/JSON_streaming#Concatenated_JSON).
- `ConcatenatedJSONStringifyStream` stringifies [Concatenated JSON](https://en.wikipedia.org/wiki/JSON_streaming#Concatenated_JSON).

Implementations have been ported from https://github.com/ayame113/jsonlines/tree/9e969323a0b09d573a4a35f0d466866bb62e1a67.

#2227 needs to be merged first. The CI should turn green when #2227 is merged into main.

### CODE REVIEW

**Review by bartlomieju — APPROVED** [review]
LGTM, too!

File: encoding/json/_stringify_test.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  options?: StringifyStreamOptions,
+) {
+  const r = new ReadableStream({
+    start(controller) {
+      for (const chunk of chunks) {
+        controller.enqueue(chunk);
+      }
+      controller.close();
+    },
+  });
```
Comment (crowlKats) [inline encoding/json/_stringify_test.ts:25]: nitpick: any reason this doesnt use `readableStreamFromIterable`?

**Review by crowlKats — APPROVED** [review]
LGTM, besides two minor nitpics

**Comment by lucacasonato** [comment]
TBH, I think these combinators are too specific. A `JSONLinesParseStream` is the same as a `LineStream` + `JSONParseStream`. For stream combinators we should provide low level primitives that users can orchestrate together themselves instead of providing one trick ponys.

I think `JSONParseStream` (that just parses each chunk as is), and `JSONStringifyStream` (which stringifies each chunk as is), are much better combinators.

**Comment by nayeemrmn** [comment]
IMO we should change this and the jsonc module to use `Json` casing as we do with `fromFileUrl()`

**Comment by kt3k** [comment]
wait we also have `URL`..

**Comment by ayame113** [comment]
Maybe the usage will be like this?
```ts
// parses JSON lines, NDJSON and JSON Text Sequences
const readable = body!
  .pipeThrough(new TextDecoderStream())
  .pipeThrough(new TextDelimiterStream(separator)) // or TextLineStream
  .pipeThrough(new JSONParseStream());

// parses Concatenated JSON
const readable = body!
  .pipeThrough(new TextDecoderStream())
  .pipeThrough(new ConcatenatedJSONParseStream());

// stringifies JSON lines, NDJSON, JSON Text Sequences and Concatenated JSON
readable
  .pipeThrough(new JSONStringifyStream({ prefix, suffix })) // default of prefix is "", default of suffix is "\n"
  .pipeThrough(new TextEncoderStream());

// or
// readable
//  .pipeThrough(new JSONStringifyStream())
//  .pipeThrough(new MappingStream((str) => `${prefix}${str}${suffix}`)) // The name is tentative
//  .pipeThrough(new TextEncoderStream());
```

I don't have time this week, so I think I'll be working on the weekend.

**Comment by kt3k** [comment]
@ayame113 The interface looks cleaner to me. Let's switch to it

**Comment by ayame113** [comment]
The current interface looks like this:

- `new JSONParseStream()` - Parse each chunk as JSON
- `new ConcatenatedJSONParseStream()` - Parse Concatenated JSON
- `new JSONStringifyStream({prefix, suffix})` - Convert each chunk to a JSON string and add a prefix and suffix

File: encoding/json/_stringify_test.ts (hunk 1 above)
Comment (ayame113) [reply encoding/json/_stringify_test.ts]: Thank you for pointing this out.

**Review by kt3k — APPROVED** [review]
LGTM again.

I'm in favor of `JSON` casing.

**Comment by kt3k** [comment]
@lucacasonato Now the APIs are more primitive. PTAL when you have time

File: encoding/README.md [hunk 2]
```diff
@@ -507,6 +508,149 @@ Serializes `object` as a YAML document.
...
+  .pipeThrough(new ConcatenatedJSONParseStream()); // parse Concatenated JSON
+
+for await (const data of readable) {
+  console.log(data);
+}
+```
+
+Use `JSONStringifyStream` to transform streaming data to
+[JSON lines](https://jsonlines.org/), [NDJSON](http://ndjson.org/),
+[NDJSON](http://ndjson.org/) or
```
Comment (kt3k) [inline encoding/README.md:589]: NDJSON seems appearing twice?

**Pull request merged**